### PR TITLE
Add UnevaluatedExpr class to handle held (unevaluated) expression.

### PR DESCRIPTION
**Pull request by Upabjojr**

### CODE REVIEW

**Comment by asmeurer** [comment]
It might be better to call this UnevaluatedExpr to match the SymPy naming scheme.

File: sympy/core/expr.py [hunk 1]
```diff
@@ -3260,6 +3260,31 @@ def _mag(x):
...
+    x*1/x
+
+    """
+
+    def __new__(cls, arg, **kwargs):
+        obj = Expr.__new__(cls, arg, **kwargs)
+        return obj
+
+    def doit(self, *args, **kwargs):
+        return self.args[0].doit(*args, **kwargs)
```
Comment (asmeurer) [inline sympy/core/expr.py]: I guess `doit(deep=False)` should just return `self.args[0]`.

**Comment by asmeurer** [comment]
This looks good. 

A question that comes up is if any methods on HeldExpr should pass through to the args, e.g., should assumptions pass through? Obviously we don't want everything to pass through because that would cause unintended evaluation.

**Comment by asmeurer** [comment]
This could also serve as a base class for things like https://github.com/sympy/sympy/issues/6835. So maybe we should leave this alone (no methods pass through), and use subclasses if we want certain methods to pass through. I think the methods we do and don't want to pass through will be dependent on the type of expression we are holding.

**Comment by Upabjojr** [comment]
> A question that comes up is if any methods on HeldExpr should pass through to the args, e.g., should assumptions pass through? Obviously we don't want everything to pass through because that would cause unintended evaluation. 

I guess assumptions should always pass through (can you think of a case where it shouldn't?).

> So maybe we should leave this alone (no methods pass through), and use subclasses if we want certain methods to pass through. I think the methods we do and don't want to pass through will be dependent on the type of expression we are holding.

If we use multiple inheritance, all methods will be inherited (which we don't want). An easy solution would be to provide a list of overloaded methods and define a `_getattr_` method.

**Comment by Upabjojr** [comment]
@asmeurer I wouldn't expand this class too much immediately. We could merge this now and then think over time how to expand this class.

**Comment by Upabjojr** [comment]
@asmeurer I wouldn't expand this class too much immediately. We could merge this now and then think over time how to expand this class.

**Comment by asmeurer** [comment]
I agree.

**Pull request merged**